go/ai/prompt/render.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prompt
+
+import (
+	"context"
+
+	"github.com/firebase/genkit/go/ai"
+)
+
+// RenderSystemPrompt renders the system prompt.
+func RenderSystemPrompt(ctx context.Context, config *Config, messages []*ai.Message, input map[string]any, raw any) ([]*ai.Message, error) {
+	var templateText string
+	var err error
+
+	if config.SystemFn != nil {
+		templateText, err = config.SystemFn(ctx, raw)
+		if err != nil {
+			return nil, err
+		}
+	} else if config.System != "" {
+		templateText = config.System
+	}
+
+	rendered, err := renderDotprompt(templateText, input, config.DefaultInput)
+	if err != nil {
+		return nil, err
+	}
+
+	if rendered != "" {
+		messages = append(messages, &ai.Message{
+			Role:    ai.RoleSystem,
+			Content: []*ai.Part{ai.NewTextPart(rendered)},
+		})
+	}
+
+	return messages, nil
+}
+
+// RenderUserPrompt renders the user prompt.
+func RenderUserPrompt(ctx context.Context, config *Config, messages []*ai.Message, input map[string]any, raw any) ([]*ai.Message, error) {
+	var templateText string
+	var err error
+
+	if config.PromptFn != nil {
+		templateText, err = config.PromptFn(ctx, raw)
+		if err != nil {
+			return nil, err
+		}
+	} else if config.Prompt != "" {
+		templateText = config.Prompt
+	}
+
+	rendered, err := renderDotprompt(templateText, input, config.DefaultInput)
+	if err != nil {
+		return nil, err
+	}
+
+	if rendered != "" {
+		messages = append(messages, &ai.Message{
+			Role:    ai.RoleUser,
+			Content: []*ai.Part{ai.NewTextPart(rendered)},
+		})
+	}
+
+	return messages, nil
+}
+
+func RenderMessages(ctx context.Context, config *Config, messages []*ai.Message, input map[string]any, raw any) ([]*ai.Message, error) {
```

Comment:
Function description here please. Here and the other functions, please briefly describe what happens (e.g. uses either the `Messages` or the `MessagesFn` part of `Config` to generate a slice of message, etc).